src/Stratis.Bitcoin/Utilities/FileStorage.cs

```diff
@@ -29,18 +29,37 @@ public FileStorage(string folderPath)
         }
 
         /// <summary>
-        /// Saves an object to a file.
+        /// Saves an object to a file, optionally keeping a backup of it.
         /// </summary>
         /// <param name="toSave">Object to save as a file.</param>
         /// <param name="fileName">Name of the file to be saved.</param>
-        public void SaveToFile(T toSave, string fileName)
+        /// <param name="saveBackupFile">A value indicating whether to save a backup of the file.</param>
+        public void SaveToFile(T toSave, string fileName, bool saveBackupFile = false)
         {
             Guard.NotEmpty(fileName, nameof(fileName));
             Guard.NotNull(toSave, nameof(toSave));
 
             string filePath = Path.Combine(this.FolderPath, fileName);
+            string tempFilePath = $"{filePath}.temp";
```

Comment:
I would get a random temp name using something like System.IO.Path.GetTempFileName(); so you reduce the chances of writing to a file that is already being locked or used in some way ...

Reply:
That's a great suggestion. I'm keen to keep the temp file in the same folder as the principal so GetTempFileName won't work for us, but I'll find a way to create unique temp file names (probably using ticks) .